crud/src/main/java/com/redhat/lightblue/assoc/QueryPlanChooser.java

```diff
@@ -184,34 +181,44 @@ private void iterateReferences(CompositeMetadata root,List<Conjunct> unassignedC
      * nodes. These are fixed assignments, and don't change from one
      * query plan to the other
      */
-    private void assignQueriesToPlanNodesAndEdges(List<Conjunct> queries,List<Conjunct> unassigned) {
+    private void assignQueriesToPlanNodesAndEdges(List<Conjunct> queries,
+                                                  List<Conjunct> unassigned) {
         Error.push("assignQueriesToPlanNodesAndEdges");
         LOGGER.debug("Assigning queries to query plan nodes and edges");
         try {
             for(Conjunct c:queries) {
-                Set<QueryPlanNode> nodes=c.getReferredNodes();
+                Set<CompositeMetadata> entities=c.getEntities();
                 LOGGER.debug("Conjunct {}",c);
-                switch(nodes.size()) {
+                switch(entities.size()) {
                 case 1:
                     // All query clauses that depend on fields from a single
                     // entity can be assigned to a query plan node. If clauses
                     // depend on multiple entities, their assignments may change
                     // based on the query plan.
                     LOGGER.debug("Conjunct has one entity");
-                    nodes.iterator().next().getData().getConjuncts().add(c);
+                    // Rewrite the query for that node
+                    QueryPlanNode node=qplan.getNode(entities.iterator().next());
+                    RewriteQuery rw=new RewriteQuery(compositeMetadata,node.getMetadata());
+                    QueryExpression q=rw.rewriteQuery(c.getClause(),c.getFieldInfo()).query;
+                    AnalyzeQuery analyzer=new AnalyzeQuery(node.getMetadata(),c.getReference());
+                    analyzer.iterate(q);
+                    node.getData().getConjuncts().add(new Conjunct(q,
+                                                                   analyzer.getFieldInfo(),
+                                                                   c.getReference()));
                     break;
 
                 case 2:
                     // There are two or more entities referred to in the conjunct
                     // This clause can be associated with an edge
```

Comment:
Yes, if a conjunct refers to two entities, and if there is an association
defined between those two entities, then the conjunct is assigned to an
edge. Otherwise, the conjunct is left unassigned, meaning that once all the
documents are retrieved, another filtering stage is necessary to filter out
any documents that don't match this particular conjunct. This happens with
something like this:

{ a:1 or b:2 }

where a belongs to antity A, and B belongs to entity B, and there is no
reference field defined between A and B. Then, we build the documents
discarding this particular predicate, and then filter out those that don't
satisfy it.

On Mon, Mar 28, 2016 at 7:37 PM, Dennis Crissman <notifications@github.com>
wrote:

> In crud/src/main/java/com/redhat/lightblue/assoc/QueryPlanChooser.java
> <https://github.com/lightblue-platform/lightblue-core/pull/608#discussion_r57661946>
> :
>
> >                  case 1:
> >                      // All query clauses that depend on fields from a single
> >                      // entity can be assigned to a query plan node. If clauses
> >                      // depend on multiple entities, their assignments may change
> >                      // based on the query plan.
> >                      LOGGER.debug("Conjunct has one entity");
> > -                    nodes.iterator().next().getData().getConjuncts().add(c);
> > +                    // Rewrite the query for that node
> > +                    QueryPlanNode node=qplan.getNode(entities.iterator().next());
> > +                    RewriteQuery rw=new RewriteQuery(compositeMetadata,node.getMetadata());
> > +                    QueryExpression q=rw.rewriteQuery(c.getClause(),c.getFieldInfo()).query;
> > +                    AnalyzeQuery analyzer=new AnalyzeQuery(node.getMetadata(),c.getReference());
> > +                    analyzer.iterate(q);
> > +                    node.getData().getConjuncts().add(new Conjunct(q,
> > +                                                                   analyzer.getFieldInfo(),
> > +                                                                   c.getReference()));
> >                      break;
> >
> >                  case 2:
> >                      // There are two or more entities referred to in the conjunct
> >                      // This clause can be associated with an edge
>
> Is the intention that if there are more than 2 entities, that the default
> option will be used?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/608/files/17418539ee2bfbf01e0fd0d96b8e9984280a6bbb#r57661946>
>